eth/proof.go

```diff
@@ -8,56 +8,98 @@ import (
 	"strings"
 	"time"
 
+	"github.com/ethereum/go-ethereum"
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/iden3/contracts-abi/rhs-storage/go/abi"
 	"github.com/iden3/go-merkletree-sql/v2"
 	"github.com/iden3/merkletree-proof"
 )
 
 type ReverseHashCli struct {
-	contract   *abi.IRHSStorage
-	txOpts     *bind.TransactOpts
-	rpcTimeout time.Duration
+	contract             *abi.IRHSStorage
+	ethClient            *ethclient.Client
+	from                 ethcommon.Address
+	signer               bind.SignerFn
+	rpcTimeout           time.Duration
+	needWaitReceipt      bool
+	txReceiptTimeout     time.Duration
+	waitReceiptCycleTime time.Duration
 }
 
-func NewReverseHashCli(contractAddress ethcommon.Address,
-	ethClient *ethclient.Client, txOpts *bind.TransactOpts,
-	defaultRPCTimeout time.Duration) (*ReverseHashCli, error) {
+type Option func(cli *ReverseHashCli) error
 
-	if ethClient == nil {
-		return nil, errors.New("ethClient is nil")
+func WithEthClient(ethClient *ethclient.Client) Option {
+	return func(cli *ReverseHashCli) error {
+		cli.ethClient = ethClient
+		return nil
 	}
+}
 
-	contract, err := abi.NewIRHSStorage(contractAddress, ethClient)
-	if err != nil {
-		return nil, fmt.Errorf("failed to instantiate a smart contract: %s", err)
+func WithRPCTimeout(timeout time.Duration) Option {
+	return func(cli *ReverseHashCli) error {
+		cli.rpcTimeout = timeout
+		return nil
 	}
+}
 
-	return &ReverseHashCli{
-		contract:   contract,
-		txOpts:     txOpts,
-		rpcTimeout: defaultRPCTimeout,
-	}, nil
+func WithNeedWaitReceipt(needWaitReceipt bool) Option {
+	return func(cli *ReverseHashCli) error {
+		cli.needWaitReceipt = needWaitReceipt
+		return nil
+	}
 }
 
-func (cli *ReverseHashCli) ctx(
-	ctx context.Context) (context.Context, context.CancelFunc) {
+func WithTxReceiptTimeout(timeout time.Duration) Option {
+	return func(cli *ReverseHashCli) error {
+		cli.txReceiptTimeout = timeout
+		return nil
+	}
+}
 
-	if ctx == nil {
-		ctx = cli.txOpts.Context
+func WithWaitReceiptCycleTime(cycleTime time.Duration) Option {
+	return func(cli *ReverseHashCli) error {
+		cli.waitReceiptCycleTime = cycleTime
+		return nil
 	}
+}
 
-	if ctx == nil {
-		ctx = context.Background()
+func NewReverseHashCli(contractAddress ethcommon.Address,
+	from ethcommon.Address, signerFn bind.SignerFn,
+	opts ...Option) (*ReverseHashCli, error) {
+
+	ethCl, err := ethclient.Dial("http://127.0.0.1:8545")
```

Comment:
probably it should not be hardcoded. I gues eth client is mandatory to init 
@olomix @AndriianChestnykh @ilya-korotya

Reply:
I agree with @vmidyllic